pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,214 @@
+				// If all tablets aren't healthy, we should bail and not perform a rolling restart.
+				r.recorder.Eventf(vts, corev1.EventTypeWarning, "RollingRestartFailed", "all tablets are not healthy")
+				return resultBuilder.Result()
+			}
+		}
+	}
+
+	podList := &corev1.PodList{}
+	listOpts := &client.ListOptions{
+		Namespace: vts.Namespace,
+		LabelSelector: apilabels.Set(map[string]string{
+			planetscalev2.ComponentLabel: planetscalev2.VttabletComponentName,
+			planetscalev2.ClusterLabel:   vts.Labels[planetscalev2.ClusterLabel],
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	scheduledTablets := false
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if !rollout.Scheduled(pod) {
+			continue
+		}
+
+		scheduledTablets = true
+		if err := r.releaseTabletPod(ctx, log, pod); err != nil {
+			r.recorder.Eventf(vts, corev1.EventTypeWarning, "RollingRestartFailed", "tablet deletion failed")
+			return resultBuilder.Error(err)
+		}
+	}
+
+	// If we have no more scheduled tablets, unschedule the shard.
+	if !scheduledTablets {
+		if err := r.unscheduleObject(ctx, vts); err != nil {
+			return resultBuilder.Error(err)
+		}
+	}
+
+	return resultBuilder.Result()
+}
+
+func (r *ReconcileVitessShard) releaseTabletPod(ctx context.Context, log *logrus.Entry, pod *corev1.Pod) error {
+	// Release the pod to be recreated with updates.
+	if err := r.releaseObject(ctx,pod); err != nil {
+		return err
+	}
+
+	// TODO: Evict pods before deleting them.
+	if err := r.client.Delete(ctx, pod); err != nil {
+		return err
+	}
+	if err := r.waitObjectUpdated(ctx, pod); err != nil {
+		return err
+	}
+	if err := r.waitPodReady(ctx, pod); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileVitessShard) waitObjectUpdated(ctx context.Context, obj runtime.Object) error {
+	// Wait for obj to have no scheduled updates.
+	for {
+		if err := regetObject(ctx, r.client, obj); err != nil {
+			// A NotFound error is ok. We wait for it to be recreated.
+			if apierrors.IsNotFound(err) {
+				time.Sleep(1 * time.Second)
+				continue
+			} else {
+				return err
+			}
+		}
+
+		if !rollout.Scheduled(objectMeta(obj)) {
+			// This is what we're waiting for.
+			return nil
+		}
+
+		// Keep waiting.
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (r *ReconcileVitessShard) waitPodReady(ctx context.Context, pod *corev1.Pod) error {
+	for {
+		if err := regetObject(ctx, r.client, pod); err != nil {
+			return err
+		}
+
+		if podIsReady(pod) {
+			// This is what we're waiting for.
+			return nil
+		}
+
+		// Keep waiting.
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (r *ReconcileVitessShard) releaseObject(ctx context.Context, obj runtime.Object) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
Unlike the CLI tool, we don't usually RetryOnConflict in controllers. Instead, we just return the error and let that trigger a requeue.

Reply:
Makes sense. Done.